### ISSUE TITLE
incorrect (Inf + Inf*im)^2.0 and other complex^float

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#24515]
In Julia 0.4–0.7, I get:
```jl
julia> (Inf + Inf*im)^2.0
0.0 + Inf*im
```
which seems wrong (it assumes that the two `Inf` components are the "same").  The correct answer is given for `^2`:
```jl
julia> (Inf + Inf*im)^2
NaN + Inf*im
```
The error seems to be due to [this line](https://github.com/JuliaLang/julia/blob/00dafe28c05e35a3a6d6230a0effe2fcf628e8d4/base/complex.jl#L641), which dates back to #2891 by @jiahao.

My first inclination is to remove the entire `p==2` special case from this `^` method.  As I commented in #24497, that seems to be a performance optimization that is probably largely superseded, since most cases that need a fast `z^2` probably use (a) an integer exponent (that calls a different method) and/or (b) a literal integer exponent (for which `z^2` is inlined as `z*z` for `Complex64` and `Complex128`).

**Comment by stevengj** [comment]
On the other hand, we have:
```
julia> m = realmax(Float64)
1.7976931348623157e308

julia> (m + m*im)^2      # incorrect
NaN + Inf*im

julia> (m + m*im)^2.0   # incorrect
0.0 + Inf*im
```
I don't think that it is worth the extra checks to deal with this case correctly, however.  You'd have to slow down *all* complex multiplications to make `(m + m*im) * (m + m*im)` give `0.0 + Inf*im`.

**Comment by stevengj** [comment]
Wow, I found some even worse bugs:
```jl
julia> (0+0im)^-3.0
0.0 + 0.0im
```
and
```jl
julia> @btime (1.0+0.0im)^1e300
ERROR: InexactError()
Stacktrace:
 [1] convert(::Type{Int64}, ::Float64) at ./float.jl:679
 [2] ^(::Complex{Float64}, ::Complex{Float64}) at ./complex.jl:593
 ...
```

**Comment by stevengj** [comment]
I'm working at a patch that fixes this, and also makes the code 60% faster for complex^complex and 120% faster for real^complex.

**Comment by stevengj** [comment]
Another bad case:
```jl
julia> Inf^(-Inf + 0.0im)
-0.0 - 0.0im
```

**Comment by stevengj** [comment]
Closed by #24570.
